ds/ringbuffer/ringbuffer.go

```diff
@@ -0,0 +1,56 @@
+package ringbuffer
+
+import (
+	"sync"
+)
+
+// RingBuffer is a thread-safe fixed buffer of elements with FIFO semantics.
+// When the buffer is full, adding a new element overwrites the oldest element.
+type RingBuffer[T any] struct {
+	buffer   []T
+	pos      int
+	capacity int
+	size     int
+	mutex    sync.RWMutex
+}
+
+// NewRingBuffer creates a new RingBuffer with a maximum size of capacity.
+func NewRingBuffer[T any](capacity int) *RingBuffer[T] {
+	return &RingBuffer[T]{
+		buffer:   make([]T, capacity),
+		capacity: capacity,
+	}
+}
+
+// Add adds an element to the buffer, overwriting the oldest element if the buffer is full.
+func (r *RingBuffer[T]) Add(element T) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	r.buffer[r.pos] = element
+	r.pos = (r.pos + 1) % r.capacity
+
+	if r.size < r.capacity {
+		r.size = r.size + 1
```

Comment:
The size never decreases? So this can never be emptied again

Reply:
Correct, size cannot decrease. We designed it after a regular buffer: you cannot decrease the size of it: the only way is to assign a new buffer and slice the old one with the element removed.